blogposts/2022/introducting-migrator-service.md

```diff
@@ -0,0 +1,149 @@
+2. We see this issue frequently in our Cloud environment, but exclusively on objects over a particular (but undetermined) size, such as our repository or code intelligence data tables.
+
+We believed that the issue occurred if and only if the set of migrations running on application startup could not finish within the parent container's health check timeout. If the migration took longer to perform than the configured timeout, Kubernetes would declare the pod dead and restart it. Killing a pod _during_ an active migration marks the database as dirty until the site administrator manually resolves the issue.
+
+And this is exactly the behavior we see.
+
+Our original solution, simply "decoupling" the database from application upgrades, targeted the Kubernetes health check as the primary problem. By performing the migrations separately from the application deployment, the migration process no longer has to complete within a configured healthcheck timeout of a downstream application. By having a separate component performing the migrations, and by moving away from our dependency on a [third-party](https://github.com/golang-migrate/migrate) migration library, we can better inspect the health of the migrator and the progress of its migration attempts.
+
+The first draft of the `migrator` service kept all other behavior the same. We explicitly kept the new migration behavior as close to the previous migration behavior as possible - just running in a different container - to avoid turning too many knobs at once. Part of the behavior we explicitly did _not_ change was how we handle concurrent migration attempts.
+
+Each migrator attempts to take an advisory lock over a shared key representing the target schema. If the lock is acquired, the current database schema version is read. If the database is marked as dirty, the migrator panics. If the database is up-to-date, the migrator can exit. Otherwise, all unapplied migration files are applied and the lock is released.
+
+After some initial testing, we discovered that the frequency of "dirty database" behaviors did not actually decrease. The solution did not address the symptoms, and we needed to do more.
+
+## A deeper contributing factor
+
+Over time, our engineering organization formed some tribal knowledge regarding ways to avoid "taking down production" with poorly defined migrations. Of particular hesitation was creating indexes on tables that can have a large number of rows. Normal index creation operations take an exclusive write lock on the table, creating the new index from the existing table data synchronously. During this time no inserts, updates, or deletes can be made to the table. You basically incur write downtime (along with a blip of failed requests in your metrics dashboards and alerting systems, and a slice of failed jobs that failed in the background) for the table, which can take tens of minutes for large tables.
```

Comment:
```suggestion
Over time, our engineering organization formed some tribal knowledge regarding ways to avoid "taking down production" with poorly defined migrations. Creating indexes on tables that can have a large number of rows caused many to hesitate. Normal index creation operations take an exclusive write lock on the table, creating the new index from the existing table data synchronously. During this time no inserts, updates, or deletes can be made to the table. You basically incur write downtime (along with a blip of failed requests in your metrics dashboards and alerting systems, and a slice of failed jobs that failed in the background) for the table, which can take tens of minutes for large tables.
```